README
Feed-PhaseCheck

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Feed::PhaseCheck

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Feed-PhaseCheck

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Feed-PhaseCheck

    CPAN Ratings
        http://cpanratings.perl.org/d/Feed-PhaseCheck

    Search CPAN
        http://search.cpan.org/dist/Feed-PhaseCheck/


LICENSE AND COPYRIGHT

Copyright (C) 2015 Maksym Kotielnikov

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Feed-PhaseCheck
{{$NEXT}}

0.07      2023-01-17 10:10:35+00:00 UTC
       Add critic test

0.06      2017-02-20 06:23:56+00:00 UTC
       switch to dzil
0.01    2015-12-23 16:57:43 +0800  
        First version, released on an unsuspecting world.



t/spots.pl
our $spots = {
    'DCFX' => {
        '1447286220' => '1.07617',
        '1447286221' => '1.07608',
        '1447286222' => '1.07608',
        '1447286224' => '1.07610',
        '1447286226' => '1.07610',
        '1447286227' => '1.07614',
        '1447286231' => '1.07617',
        '1447286232' => '1.07617',
        '1447286235' => '1.07617',
        '1447286246' => '1.07617',
        '1447286247' => '1.07617',
        '1447286254' => '1.07614',
        '1447286255' => '1.07608',
        '1447286256' => '1.07610',
        '1447286257' => '1.07611',
        '1447286258' => '1.07616',
        '1447286262' => '1.07617',
        '1447286263' => '1.07617',
        '1447286268' => '1.07618',
        '1447286269' => '1.07621',
        '1447286271' => '1.07621',
        '1447286274' => '1.07614',
        '1447286276' => '1.07614',
        '1447286289' => '1.07612',
        '1447286295' => '1.07611',
        '1447286304' => '1.07609',
        '1447286305' => '1.07610',
        '1447286309' => '1.07611',
        '1447286314' => '1.07610',
        '1447286317' => '1.07609',
        '1447286318' => '1.07610',
        '1447286320' => '1.07623',
        '1447286321' => '1.07631',
        '1447286322' => '1.07631',
        '1447286324' => '1.07630',
        '1447286331' => '1.07631',
        '1447286332' => '1.07632',
        '1447286333' => '1.07631',
        '1447286338' => '1.07631',
        '1447286339' => '1.07631',
        '1447286340' => '1.07631',
        '1447286347' => '1.07632',
        '1447286348' => '1.07633',
        '1447286349' => '1.07637',
        '1447286355' => '1.07637',
        '1447286356' => '1.07638',
        '1447286359' => '1.07638',
        '1447286360' => '1.07637',
        '1447286362' => '1.07636',
        '1447286366' => '1.07637',
        '1447286368' => '1.07636',
        '1447286369' => '1.07635',
        '1447286371' => '1.07635',
        '1447286372' => '1.07635',
        '1447286375' => '1.07635',
        '1447286380' => '1.07635',
        '1447286382' => '1.07635',
        '1447286385' => '1.07636',
        '1447286398' => '1.07635',
        '1447286399' => '1.07635'
    },
    'FXCM' => {
        '1447286220' => '1.07618',
        '1447286222' => '1.076135',
        '1447286223' => '1.0761125',
        '1447286224' => '1.07609',
        '1447286225' => '1.076095',
        '1447286226' => '1.07610',
        '1447286227' => '1.07615',
        '1447286228' => '1.07616',
        '1447286229' => '1.07616333333333',
        '1447286230' => '1.07616666666667',
        '1447286231' => '1.07617',
        '1447286232' => '1.07617',
        '1447286233' => '1.076175',
        '1447286234' => '1.07618',
        '1447286235' => '1.07615',
        '1447286236' => '1.07617',
        '1447286237' => '1.07616888888889',
        '1447286238' => '1.07616777777778',
        '1447286239' => '1.07616666666667',
        '1447286240' => '1.07616555555556',
        '1447286241' => '1.07616444444444',
        '1447286242' => '1.07616333333333',
        '1447286243' => '1.07616222222222',
        '1447286244' => '1.07616111111111',
        '1447286245' => '1.07616',
        '1447286246' => '1.07615888888889',
        '1447286247' => '1.07615777777778',
        '1447286248' => '1.07615666666667',
        '1447286249' => '1.07615555555556',
        '1447286250' => '1.07615444444444',
        '1447286251' => '1.07615333333333',
        '1447286252' => '1.07615222222222',
        '1447286253' => '1.07615111111111',
        '1447286254' => '1.07615',
        '1447286255' => '1.076115',
        '1447286256' => '1.07608',
        '1447286257' => '1.07610',
        '1447286258' => '1.07618',
        '1447286259' => '1.07618',
        '1447286260' => '1.07618',
        '1447286261' => '1.07618',
        '1447286262' => '1.07618',
        '1447286263' => '1.07618',
        '1447286264' => '1.07618',
        '1447286265' => '1.07618',
        '1447286266' => '1.07618',
        '1447286267' => '1.07618',
        '1447286268' => '1.07618',
        '1447286269' => '1.07620',
        '1447286270' => '1.07621',
        '1447286271' => '1.07622',
        '1447286272' => '1.076195',
        '1447286273' => '1.07617',
        '1447286274' => '1.07617',
        '1447286275' => '1.07617',
        '1447286276' => '1.07617',
        '1447286277' => '1.07616',
        '1447286278' => '1.07617',
        '1447286279' => '1.07617',
        '1447286280' => '1.076165',
        '1447286281' => '1.07616',
        '1447286282' => '1.07615',
        '1447286283' => '1.076155',
        '1447286284' => '1.07616',
        '1447286285' => '1.076155',
        '1447286286' => '1.07615',
        '1447286287' => '1.076155',
        '1447286288' => '1.07616',
        '1447286289' => '1.07614',
        '1447286290' => '1.07612',
        '1447286291' => '1.07612',
        '1447286292' => '1.07612',
        '1447286293' => '1.07612',
        '1447286294' => '1.07612',
        '1447286295' => '1.07612',
        '1447286296' => '1.0761175',
        '1447286297' => '1.076115',
        '1447286298' => '1.0761125',
        '1447286299' => '1.07611',
        '1447286300' => '1.07611',
        '1447286301' => '1.07611',
        '1447286302' => '1.07611',
        '1447286303' => '1.07611',
        '1447286304' => '1.07610',
        '1447286305' => '1.0761',
        '1447286306' => '1.0761',
        '1447286307' => '1.0761',
        '1447286308' => '1.0761',
        '1447286309' => '1.0761',
        '1447286310' => '1.0761',
        '1447286311' => '1.0761',
        '1447286312' => '1.07610',
        '1447286313' => '1.0761',
        '1447286314' => '1.07610',
        '1447286315' => '1.07609666666667',
        '1447286316' => '1.07609333333333',
        '1447286317' => '1.07609',
        '1447286318' => '1.07610',
        '1447286319' => '1.07623',
        '1447286320' => '1.07630',
        '1447286321' => '1.07629',
        '1447286322' => '1.07629272727273',
        '1447286323' => '1.07629545454545',
        '1447286324' => '1.07629818181818',
        '1447286325' => '1.07630090909091',
        '1447286326' => '1.07630363636364',
        '1447286327' => '1.07630636363636',
        '1447286328' => '1.07630909090909',
        '1447286329' => '1.07631181818182',
        '1447286330' => '1.07631454545455',
        '1447286331' => '1.07631727272727',
        '1447286332' => '1.07632',
        '1447286333' => '1.07631',
        '1447286334' => '1.07631',
        '1447286335' => '1.07631',
        '1447286336' => '1.07631',
        '1447286337' => '1.07631',
        '1447286338' => '1.07631',
        '1447286339' => '1.07631',
        '1447286340' => '1.07631',
        '1447286341' => '1.07631142857143',
        '1447286342' => '1.07631285714286',
        '1447286343' => '1.07631428571429',
        '1447286344' => '1.07631571428571',
        '1447286345' => '1.07631714285714',
        '1447286346' => '1.07631857142857',
        '1447286347' => '1.07632',
        '1447286348' => '1.07633',
        '1447286349' => '1.07638',
        '1447286350' => '1.07638',
        '1447286351' => '1.07638',
        '1447286352' => '1.07638',
        '1447286353' => '1.07638',
        '1447286354' => '1.07638',
        '1447286355' => '1.07638',
        '1447286356' => '1.07638',
        '1447286357' => '1.07638',
        '1447286358' => '1.07638',
        '1447286359' => '1.07638',
        '1447286360' => '1.07637',
        '1447286361' => '1.076375',
        '1447286362' => '1.07638',
        '1447286363' => '1.07637',
        '1447286364' => '1.07637333333333',
        '1447286365' => '1.07637666666667',
        '1447286366' => '1.07638',
        '1447286367' => '1.076375',
        '1447286368' => '1.07637',
        '1447286369' => '1.07637',
        '1447286370' => '1.07636666666667',
        '1447286371' => '1.07636333333333',
        '1447286372' => '1.07636',
        '1447286373' => '1.07636',
        '1447286374' => '1.07636',
        '1447286375' => '1.07635',
        '1447286376' => '1.07635',
        '1447286377' => '1.07635',
        '1447286378' => '1.07635',
        '1447286379' => '1.07635',
        '1447286380' => '1.07635',
        '1447286381' => '1.07635',
        '1447286382' => '1.07635',
        '1447286383' => '1.07636',
        '1447286384' => '1.07637',
        '1447286385' => '1.07636',
        '1447286386' => '1.07635923076923',
        '1447286387' => '1.07635846153846',
        '1447286388' => '1.07635769230769',
        '1447286389' => '1.07635692307692',
        '1447286390' => '1.07635615384615',
        '1447286391' => '1.07635538461538',
        '1447286392' => '1.07635461538462',
        '1447286393' => '1.07635384615385',
        '1447286394' => '1.07635307692308',
        '1447286395' => '1.07635230769231',
        '1447286396' => '1.07635153846154',
        '1447286397' => '1.07635076923077',
        '1447286398' => '1.07635',
        '1447286399' => '1.07635'
    }};



t/main-hash.pl
our $main = {
    "1451276629" => "1.09765",
    "1451276630" => "1.09764916666667",
    "1451276631" => "1.09764833333333",
    "1451276632" => "1.0976475",
    "1451276633" => "1.09764666666667",
    "1451276634" => "1.09764583333333",
    "1451276635" => "1.097645",
    "1451276636" => "1.09764416666667",
    "1451276637" => "1.09764333333333",
    "1451276638" => "1.0976425",
    "1451276639" => "1.09764166666667",
    "1451276640" => "1.09764083333333",
    "1451276641" => "1.09764",
    "1451276642" => "1.097635",
    "1451276643" => "1.09764",
    "1451276644" => "1.097645",
    "1451276645" => "1.0976445",
    "1451276646" => "1.097644",
    "1451276647" => "1.0976435",
    "1451276648" => "1.097643",
    "1451276649" => "1.0976425",
    "1451276650" => "1.097642",
    "1451276651" => "1.0976415",
    "1451276652" => "1.097641",
    "1451276653" => "1.0976405",
    "1451276654" => "1.09764",
    "1451276655" => "1.097645",
    "1451276656" => "1.097645",
    "1451276657" => "1.097645",
    "1451276658" => "1.09764428571429",
    "1451276659" => "1.09764357142857",
    "1451276660" => "1.09764285714286",
    "1451276661" => "1.09764214285714",
    "1451276662" => "1.09764142857143",
    "1451276663" => "1.09764071428571",
    "1451276664" => "1.09764",
    "1451276665" => "1.09763928571429",
    "1451276666" => "1.09763857142857",
    "1451276667" => "1.09763785714286",
    "1451276668" => "1.09763714285714",
    "1451276669" => "1.09763642857143",
    "1451276670" => "1.09763571428571",
    "1451276671" => "1.097635",
    "1451276672" => "1.09764",
    "1451276673" => "1.097645",
    "1451276674" => "1.09764",
    "1451276675" => "1.097645",
    "1451276676" => "1.09765",
    "1451276677" => "1.097655",
    "1451276678" => "1.0976546",
    "1451276679" => "1.0976542",
    "1451276680" => "1.0976538",
    "1451276681" => "1.0976534",
    "1451276682" => "1.097653",
    "1451276683" => "1.0976526",
    "1451276684" => "1.0976522",
    "1451276685" => "1.0976518",
    "1451276686" => "1.0976514",
    "1451276687" => "1.097651",
    "1451276688" => "1.0976506",
    "1451276689" => "1.0976502",
    "1451276690" => "1.0976498",
    "1451276691" => "1.0976494",
    "1451276692" => "1.097649",
    "1451276693" => "1.0976486",
    "1451276694" => "1.0976482",
    "1451276695" => "1.0976478",
    "1451276696" => "1.0976474",
    "1451276697" => "1.097647",
    "1451276698" => "1.0976466",
    "1451276699" => "1.0976462",
    "1451276700" => "1.0976458",
    "1451276701" => "1.0976454",
    "1451276702" => "1.097645",
    "1451276703" => "1.09764447368421",
    "1451276704" => "1.09764394736842",
    "1451276705" => "1.09764342105263",
    "1451276706" => "1.09764289473684",
    "1451276707" => "1.09764236842105",
    "1451276708" => "1.09764184210526",
    "1451276709" => "1.09764131578947",
    "1451276710" => "1.09764078947368",
    "1451276711" => "1.09764026315789",
    "1451276712" => "1.09763973684211",
    "1451276713" => "1.09763921052632",
    "1451276714" => "1.09763868421053",
    "1451276715" => "1.09763815789474",
    "1451276716" => "1.09763763157895",
    "1451276717" => "1.09763710526316",
    "1451276718" => "1.09763657894737",
    "1451276719" => "1.09763605263158",
    "1451276720" => "1.09763552631579",
    "1451276721" => "1.097635",
    "1451276722" => "1.0976",
    "1451276723" => "1.09759",
    "1451276724" => "1.09759",
    "1451276725" => "1.09759",
    "1451276726" => "1.09759",
    "1451276727" => "1.09759",
    "1451276728" => "1.09759",
    "1451276729" => "1.09759",
    "1451276730" => "1.09759",
    "1451276731" => "1.09759",
    "1451276732" => "1.09759",
    "1451276733" => "1.09759",
    "1451276734" => "1.09759",
    "1451276735" => "1.09759",
    "1451276736" => "1.09759",
    "1451276737" => "1.09759",
    "1451276738" => "1.09759",
    "1451276739" => "1.09759",
    "1451276740" => "1.09759",
    "1451276741" => "1.09759",
    "1451276742" => "1.09759",
    "1451276743" => "1.09759",
    "1451276744" => "1.09759",
    "1451276745" => "1.09759",
    "1451276746" => "1.09759",
    "1451276747" => "1.09759",
    "1451276748" => "1.09759",
    "1451276749" => "1.09759",
    "1451276750" => "1.09759",
    "1451276751" => "1.09759",
    "1451276752" => "1.09759",
    "1451276753" => "1.09759",
    "1451276754" => "1.09759",
    "1451276755" => "1.09759166666667",
    "1451276756" => "1.09759333333333",
    "1451276757" => "1.097595",
    "1451276758" => "1.0975925",
    "1451276759" => "1.09759",
    "1451276760" => "1.09759333333333",
    "1451276761" => "1.09759666666667",
    "1451276762" => "1.0976",
    "1451276763" => "1.097595",
    "1451276764" => "1.097595",
    "1451276765" => "1.097595",
    "1451276766" => "1.097595",
    "1451276767" => "1.097595",
    "1451276768" => "1.097595",
    "1451276769" => "1.097595",
    "1451276770" => "1.097595",
    "1451276771" => "1.097595",
    "1451276772" => "1.097595",
    "1451276773" => "1.097595",
    "1451276774" => "1.097595",
    "1451276775" => "1.097595",
    "1451276776" => "1.097595",
    "1451276777" => "1.097595",
    "1451276778" => "1.097595",
    "1451276779" => "1.097595",
    "1451276780" => "1.097595",
    "1451276781" => "1.097595",
    "1451276782" => "1.097595",
    "1451276783" => "1.097595",
    "1451276784" => "1.097595",
    "1451276785" => "1.097595",
    "1451276786" => "1.097595",
    "1451276787" => "1.097595",
    "1451276788" => "1.097595",
    "1451276789" => "1.097595",
    "1451276790" => "1.097595",
    "1451276791" => "1.097595",
    "1451276792" => "1.097595"
};


t/sample-hash.pl
our $sample = {
    "1451276654" => "1.097655",
    "1451276655" => "1.09765",
    "1451276656" => "1.097655",
    "1451276657" => "1.097625",
    "1451276671" => "1.09763",
    "1451276672" => "1.097625",
    "1451276674" => "1.09764",
    "1451276677" => "1.09765",
    "1451276678" => "1.097655",
    "1451276697" => "1.097655",
    "1451276702" => "1.09765",
    "1451276703" => "1.09764",
    "1451276717" => "1.097625",
    "1451276721" => "1.09763",
    "1451276722" => "1.097585",
    "1451276723" => "1.097595",
    "1451276724" => "1.0976",
    "1451276729" => "1.097595",
    "1451276730" => "1.0976",
    "1451276754" => "1.097585",
    "1451276757" => "1.09759",
    "1451276759" => "1.097585",
    "1451276760" => "1.09759",
    "1451276762" => "1.097595",
    "1451276763" => "1.0976",
    "1451276764" => "1.097595"
};

Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.024.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 7.64;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Finds the relative time delay between two feed segments.",
  "AUTHOR" => "binary.com <BINARY\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "7.64"
  },
  "DISTNAME" => "Feed-PhaseCheck",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Feed::PhaseCheck",
  "PREREQ_PM" => {
    "Exporter" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "POSIX" => 0,
    "Test::CheckDeps" => "0.010",
    "Test::More" => "0.94"
  },
  "VERSION" => "0.08",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "POSIX" => 0,
  "Test::CheckDeps" => "0.010",
  "Test::More" => "0.94",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


phase_check.pl
use strict;
use warnings;
use lib 'lib';
use lib 'lib/Feed-PhaseCheck/lib';

use BinaryPhaseCheck;

my ($path, $long, $short, $symbol, $interval, $where);
for (my $i = 0; $i < scalar @ARGV; $i++) {
    if ($ARGV[$i] eq '-p') {
        $path = $ARGV[$i + 1];
    } elsif ($ARGV[$i] eq '-l') {
        $long = $ARGV[$i + 1];
    } elsif ($ARGV[$i] eq '-s') {
        $short = $ARGV[$i + 1];
    } elsif ($ARGV[$i] eq '-u') {
        $symbol = $ARGV[$i + 1];
    } elsif ($ARGV[$i] eq '-i') {
        $interval = $ARGV[$i + 1];
    } elsif ($ARGV[$i] eq '-w') {
        $where = $ARGV[$i + 1];
    } elsif ($ARGV[$i] eq '-h') {
        print qq~	-p = path to feed folder
    -w = path to folder where to save the result
	-l = long period in seconds
	-s = short period in seconds
	-u = underlying name
	-i = calculate delay every i seconds
	-h = this manual
~;
        exit;
    }
}

my @errors;
if (!$path || !(-e $path)) {
    push @errors, "Incorrect path to feed folder!";
}

if ($where && !(-e $where)) {
    push @errors, "Incorrect path to result folder is provided!";
}

unless ($long && $long =~ /^[0-9]+$/ && $long >= 2 * 60 && $long <= 20 * 60) {
    push @errors, "Long period should be between 2 and 20 minutes";
}

unless ($short && $short =~ /^[0-9]+$/ && $short >= 2 * 60 && $short < $long) {
    push @errors, "Short period should be between 2 minutes and long period";
}

unless ($symbol && $symbol =~ /^[0-9a-z]+$/i) {
    push @errors, "Incorrect underlying name is provided";
}

unless ($interval && $interval =~ /^[0-9]+$/ && $interval > 0) {
    push @errors, "Incorrect interval is provided";
}

if (scalar @errors) {
    print join("\n", @errors) . "\n";
    exit;
}
my $ph = BinaryPhaseCheck->new({
        symbol      => $symbol,
        short       => $short,
        long        => $long,
        path        => $path,
        result_path => $where,
        period      => $interval,
        test        => 1
    });
$ph->calculate;
use Data::Dumper;
print Dumper $ph;



BinaryPhaseCheck.pm
use strict;
use warnings;

package BinaryPhaseCheck;
use base 'Feed::PhaseCheck';
use DataDog::DogStatsd::Helper qw(stats_gauge);

sub set {
    my ($self, $params) = @_;
    if (ref $params eq 'HASH') {
        $params->{ref} = 'FXCM';
        $self->SUPER::set($params);
        foreach (keys %$params) {
            if ($_ eq 'path') {
                $self->setPath($params->{$_});
            } elsif ($_ eq 'result_path') {
                $self->setResultPath($params->{$_});
            } elsif ($_ eq 'symbol') {
                $self->setSymbol($params->{$_});
            } elsif ($_ eq 'period') {
                $self->setPeriod($params->{$_});
            } elsif ($_ eq 'long') {
                $self->setLong($params->{$_});
            } elsif ($_ eq 'test') {
                $self->{test} = 1;
            }
            delete $params->{$_};
        }
    }
}

sub setPath {
    my ($self, $path) = @_;
    if (-e $path) {
        $self->{path} = $path;
        return 1;
    }
}

sub setResultPath {
    my ($self, $path) = @_;
    if (-e $path) {
        $self->{result_path} = $path;
        return 1;
    }
}

sub setSymbol {
    my ($self, $symbol) = @_;
    if ($symbol) {
        $self->{symbol} = $symbol;
        return 1;
    }
}

sub setPeriod {
    my ($self, $period) = @_;
    if ($period) {
        $self->{period} = $period;
        return 1;
    }
}

sub setLong {
    my ($self, $long) = @_;
    if ($long) {
        $self->{long} = $long;
        return 1;
    }
}

sub run {
    my $self = shift;

    unless ($self->{period}) {
        die("Timeout should be provided!\n");
    }

    while (1) {
        $self->calculate;
        use Data::Dumper;
        print Dumper $self;
        $self->save_to_file;
        $self->send_to_datadog;
        sleep $self->{period};
    }
}

sub calculate {
    my $self  = shift;
    my $spots = $self->tail_idata;
    $self->SUPER::set_spots($spots);
    $self->SUPER::calculate_errors;
}

sub send_to_datadog {
    my $self = shift;

    my $last_epoch = $self->{last_ref_epoch} || '';

    foreach (keys %{$self->{min_errors}}) {
        if ($self->{ticks_number}->{$_}) {
            my $error = $self->{errors}->{$_}->{0} - $self->{min_errors}->{$_}->{error};
            stats_gauge('phase_delay', $self->{min_errors}->{$_}->{delay}, {tags => ['tag:' . $_]});
            stats_gauge('phase_error', $error, {tags => ['tag:' . $_]});
        }
    }
}

sub save_to_file {
    my $self = shift;

    my @errors;
    unless ($self->{result_path}) {
        push @errors, 'Path for saving result should be provided!';
    }
    unless ($self->{symbol}) {
        push @errors, 'Underlying should be provided!';
    }

    if (scalar @errors) {
        print join("\n", @errors) . "\n";
        die();
    }

    my $path = "$self->{result_path}/$self->{symbol}.csv";

    open(my $fh, ">>", $path) or die("Can not create file!\n");

    my $last_epoch = $self->{last_ref_epoch} || '';
    foreach (sort { $a cmp $b } keys %{$self->{min_errors}}) {
        if ($self->{ticks_number}->{$_}) {
            my $error = $self->{errors}->{$_}->{0} - $self->{min_errors}->{$_}->{error};
            print $fh "$_,$self->{min_errors}->{$_}->{delay},$error,$self->{ticks_number}->{$_},$last_epoch\n";
        }
    }

    close $fh;
}

sub tail_idata {
    my $self = shift;

    my @errors;
    unless ($self->{symbol}) {
        push @errors, "Underlying is needed!";
    }
    unless ($self->{path}) {
        push @errors, "Path is needed!";
    }
    unless ($self->{long}) {
        push @errors, "Long interval is needed!";
    }

    if (scalar @errors) {
        print join("\n", @errors) . "\n";
        die();
    }

    my $interval = $self->{long};

    my %spots;
    my $last_epoch;
    my $spots = {};

    if ($self->{test}) {
        use Time::Local;
        my $path = "/home/EUR-A0-FX3.log";

        if (!(-f $path)) {
            return {};
        }

        my $bw;
        if (-e '/usr/bin/tac') {
            open $bw, "-|", "tac", $path;
        } else {
            open $bw, "-|", "tail", "-r", $path;
        }

        while (my $line = <$bw>) {
            chomp $line;
            my @fields = split /\,/, $line;

            my ($year, $month, $day, $hours, $min, $sec) = split(/[\/: ]/, $fields[1]);
            my $epoch = Time::Local::timegm($sec, $min, $hours, $day, $month - 1, $year);
            $last_epoch = $epoch if !$last_epoch;
            last if $epoch + $interval < $last_epoch;
            $spots->{$fields[4]} = {} if !$spots->{$fields[4]};
            $spots->{$fields[4]}->{$epoch} = ($fields[5] + $fields[6]) / 2;
        }
        close $bw;
    } else {
        my $date = $self->_get_current_date;
        my $path = "$self->{path}/idata/$self->{symbol}/$date-fullfeed.csv";

        if (!(-f $path)) {
            return {};
        }

        my $bw;
        if (-e '/usr/bin/tac') {
            open $bw, "-|", "tac", $path;
        } else {
            open $bw, "-|", "tail", "-r", $path;
        }

        while (my $line = <$bw>) {
            chomp $line;
            my @fields = split /\,/, $line;
            $last_epoch = $fields[0] if !$last_epoch;
            last if $fields[0] + $interval < $last_epoch;
            next if $fields[6] && $fields[6] =~ /BADSRC/;
            $spots->{$fields[5]} = {} if !$spots->{$fields[5]};
            $spots->{$fields[5]}->{$fields[0]} = $fields[4];
        }
        close $bw;
    }
    return $spots;
}

sub _get_current_date {
    my $self = shift;
    my ($sec, $min, $hour, $day, $month, $year) = gmtime(1447286399);
    my $date = "$day-" . $self->_month_to_name($month) . "-" . ($year - 100);
    return $date;
}

sub _month_to_name {
    my $self   = shift;
    my $number = shift;
    my %m      = (
        0  => 'Jan',
        1  => 'Feb',
        2  => 'Mar',
        3  => 'Apr',
        4  => 'May',
        5  => 'Jun',
        6  => 'Jul',
        7  => 'Aug',
        8  => 'Sep',
        9  => 'Oct',
        10 => 'Nov',
        11 => 'Dec'
    );
    return $m{$number};
}

1;


lib/Feed/PhaseCheck.pm
package Feed::PhaseCheck;

use 5.006;
use strict;
use warnings;

use Exporter qw(import);

our @EXPORT_OK = qw(compare_feeds);

# ABSTRACT: Finds the relative time delay between two feed segments.

=head1 NAME

Feed::PhaseCheck

Finds the relative time delay between two feed segments.

Accomplished by shifting one feed relative to the other and then computing the error (absolute difference).

The shift that yields the lowest error corresponds to the relative delay between he two input feeds.

The output consists of the delay found, and the error in delayed point.

=cut

our $VERSION = '0.08';

=head1 SYNOPSIS

    use Feed::PhaseCheck qw(compare_feeds);
    my $sample = {
        "1451276654" => "1.097655",
        "1451276655" => "1.09765",
        #...
        "1451276763" => "1.0976",
        "1451276764" => "1.097595"
    };
    my $compare_to = {
        "1451276629" => "1.09765",
        "1451276630" => "1.09764916666667",
        #...
        "1451276791" => "1.097595",
        "1451276792" => "1.097595"
    };
    my $max_delay_check = 30;    # seconds
    my ($errors,$delay_with_min_err) = compare_feeds($sample,$compare_to,$max_delay_check);

=cut

=head1 METHODS

=head2 compare_feeds

=cut

sub compare_feeds {
    my $sample          = shift;
    my $main_data       = shift;
    my $max_delay_check = shift || 0;

    if ($max_delay_check !~ /^\d+$/) {
        return;
    }

    if (ref $sample ne 'HASH' || scalar keys %$sample < 2) {
        return;
    }

    if (ref $main_data ne 'HASH' || scalar keys %$main_data < 2) {
        return;
    }

    my @main_epoches = sort keys %$main_data;
    foreach (@main_epoches) {
        if (int($_) != $_ || abs($main_data->{$_}) != $main_data->{$_}) {
            return;
        }
    }

    my @sample_epoches = sort keys %$sample;
    foreach (@sample_epoches) {
        if (int($_) != $_ || abs($sample->{$_}) != $sample->{$_}) {
            return;
        }
    }

    if ($sample_epoches[0] < $main_epoches[0] || $sample_epoches[-1] > $main_epoches[-1]) {
        return;
    }
    my %main_data = %$main_data;
    my %error     = ();
    my ($min_error, $delay_for_min_error);
    my $delay1 = $sample_epoches[0] - $main_epoches[0] < $max_delay_check   ? $sample_epoches[0] - $main_epoches[0]   : $max_delay_check;
    my $delay2 = $main_epoches[-1] - $sample_epoches[-1] < $max_delay_check ? $main_epoches[-1] - $sample_epoches[-1] : $max_delay_check;
    for (my $delay = -$delay1; $delay <= $delay2; $delay++) {
        $error{$delay} = 0;
        foreach my $epoch (@sample_epoches) {
            my $sample_epoch = $epoch + $delay;
            if (!defined $main_data{$sample_epoch}) {
                for (my $i = 1; $i < scalar keys @main_epoches; $i++) {
                    if ($main_epoches[$i] > $sample_epoch) {
                        $main_data{$sample_epoch} = _interpolate(
                            $main_epoches[$i - 1],
                            $main_data{$main_epoches[$i - 1]},
                            $main_epoches[$i], $main_data{$main_epoches[$i]},
                            $sample_epoch
                        );
                        last;
                    }
                }
            }
            $error{$delay} += ($main_data{$sample_epoch} - $sample->{$epoch})**2;
        }
        if (!defined $min_error || $error{$delay} < $min_error) {
            $min_error           = $error{$delay};
            $delay_for_min_error = $delay;
        }
        # $error{$delay} =~ s/(\d{8}).+?e/$1e/;
    }

    return (\%error, $delay_for_min_error);
}

sub _interpolate {
    my ($x1, $y1, $x2, $y2, $x) = @_;
    my $y = $y1 + ($x - $x1) * ($y2 - $y1) / ($x2 - $x1);
    return $y;
}

=head1 AUTHOR

Maksym Kotielnikov, C<< <maksym at binary.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-feed-phasecheck at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Feed-PhaseCheck>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Feed::PhaseCheck


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Feed-PhaseCheck>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Feed-PhaseCheck>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Feed-PhaseCheck>

=item * Search CPAN

L<http://search.cpan.org/dist/Feed-PhaseCheck/>

=back


=head1 ACKNOWLEDGEMENTS



=cut

1;    # End of Feed::PhaseCheck
